### PR TITLE
fix(provider): fix npm provider to update package-lock.json and npm-shrinkwrap.json if they exist

### LINKED ISSUE #804: Expected Version Provider Behavior - npm should update package.json, package-lock.json if present, and npm-shrinkwrap.json if present.
### Description

`NomProvider` currently only updates package.json. However, it is common practice to check both package.json and package-lock.json into git / scm, thus both files should be updated simultaneously.

### Possible Solution

Using `NpmProvider` should by default update both package.json and package-lock.json, with the option of skipping package-lock.json.

Another option would be to have `JsonProvider` update multiple files using with a unique JSON expression per file. This would allow developers to implement on their own. Configuration could look something like:

```toml
[tool.commitizen]
version_provider = "json"
version_files = [
    "package.json:.version",
    "package-lock.json:.version",
    "package-lock.json:.package.''.version"
]
```

### Additional context

See issue #659 

### Additional context

_No response_

UPDATED: Fixed the JSON expressions to match what should really be matched.

### PR DESCRIPTION
**Pull request by davidlday**
closes #804

## Description

Added new version provider `npm2` which uses `jsonpath_ng` to update version in

* package.json: `$.version`
* package-lock.json (if it exists): `$.version`, `$.packages.''.version`
* npm-shrinkwrap.json (if it exists): `$.version`, `$.packages.''.version`

`package-lock.json` and `npm-shrinkwrap.json` may or may not exist.

## Checklist

- [x] Add test cases to all the changes you introduce
- [x] Run `./scripts/format` and `./scripts/test` locally to ensure this change passes linter check and test
- [x] Test the changes on the local machine manually
- [x] Update the documentation for the changes

## Expected behavior

When using `npm2` as the version provider, the following files will have the project's version number updated appropriately:

* `package.json` (required)
* `package-lock.json` (optional)
* `npm-shrinkwrap.json` (optional)

## Steps to Test This Pull Request

Test cases added to `tests/test_version_providers.py`.

### CODE REVIEW

File: commitizen/providers.py [hunk 1]
```diff
@@ -1,6 +1,7 @@
 from __future__ import annotations
 
 import json
+import jsonpath_ng  # type: ignore
```
Comment (Lee-W) [inline commitizen/providers.py:4]: May I know why do we need `#type: ignore` here

File: commitizen/providers.py [hunk 2]
```diff
@@ -166,6 +167,87 @@ class NpmProvider(JsonProvider):
     filename = "package.json"
 
 
+class Npm2Provider(VersionProvider):
```
Comment (Lee-W) [inline commitizen/providers.py:170]: As mentioned in the issue, I'd suggest we update the `NpmProvider` directly, but I'm not a JavaScript users. Would love to see how @woile thinks.

File: commitizen/providers.py (hunk 1 above)
Comment (Lee-W) [reply commitizen/providers.py:4]: Also, may I know why do we need to install this package?

**Review by Lee-W — COMMENTED** [review]
Just left a few comments 🙂

File: commitizen/providers.py (hunk 2 above)
Comment (woile) [reply commitizen/providers.py:170]: I think so as well, if `package-lock.json` and or `npm-shrinkwrap.json` exists they should be updated.

File: commitizen/providers.py (hunk 1 above)
Comment (woile) [reply commitizen/providers.py:4]: I'd also like to know, is it not possible to update it accessing like a json?

Comment (davidlday) [reply commitizen/providers.py:4]: The problem came from trying to access the correct element in "packages". Here's the sample JSON:

```json
{
  "name": "whatever",
  "version": "0.1.0",
  "lockfileVersion": 2,
  "requires": true,
  "packages": {
    "": {
      "name": "whatever",
      "version": "0.1.0"
    },
    "someotherpackage": {
      "version": "0.1.0"
    }
  }
}
```

The json path expression is `$.packages.''.version`, but I couldn't get that to work via the json object directly. 

And the `# type: ignore` is there b/c the library doesn't appear to have type hints, nor is there anything available for it on typeshed.

File: commitizen/providers.py (hunk 2 above)
Comment (davidlday) [reply commitizen/providers.py:170]: Makes sense. I'll go back and update the existing NpmProvider  and take out the Npm2Provider.

Comment (davidlday) [reply commitizen/providers.py:170]: Should I change the PR from feat to fix then?

File: commitizen/providers.py (hunk 1 above)
Comment (davidlday) [reply commitizen/providers.py:4]: Okay - so not sure why I was struggling with using the json.loads and updated the dict, but I have that working now. Will remove the dependency on jsonpath_ng.

File: docs/config.md [hunk 3]
```diff
@@ -293,15 +293,16 @@ But you can use any `commitizen.provider` entrypoint as value for `version_provi
...
+| `poetry`     | Get and set version from `pyproject.toml` `tool.poetry.version` field                                                                                                                                                             |
+| `cargo`      | Get and set version from `Cargo.toml` `project.version` field                                                                                                                                                                     |
+| `npm`        | Get and set version from `package.json` `project.version` field                                                                                                                                                                   |
+| `npm2`       | Get and set version from `package.json` `$.version` field, `package-lock.json` `$.version,$.packages.''.version` fields if the file exists, and `npm-shrinkwrap.json` `$.version,$.packages.''.version` fields if the file exists |
```
Comment (Lee-W) [inline docs/config.md:304]: The doc here needs to be updated

File: commitizen/providers.py [hunk 4]
```diff
@@ -160,10 +160,77 @@ def set(self, document: dict[str, Any], version: str):
...
+        return Path() / self.shrinkwrap_filename
+
+    def get_version(self) -> str:
+        """
+        Get the current version from package.json
+        """
+        package_document = json.loads(self.package_file.read_text())
+        return self.get_package_version(package_document)
+
+    def set_version(self, version: str):
```
Comment (Lee-W) [inline commitizen/providers.py:190]:
```suggestion
    def set_version(self, version: str) -> None:
```

File: tests/test_version_providers.py [hunk 5]
```diff
@@ -225,3 +225,125 @@ def test_scm_provider_default_without_commits_and_tags(config: BaseConfig):
...
+    config: BaseConfig,
+    chdir: Path,
+    pkg_content: str,
+    pkg_expected: str,
+    pkg_lock_content: str,
+    pkg_lock_expected: str,
+    pkg_shrinkwrap_content: str,
+    pkg_shrinkwrap_expected: str,
+):
+    pkg = chdir / "package.json"
```
Comment (Lee-W) [inline tests/test_version_providers.py:330]:
```suggestion
@pytest.mark.parametrize("pkg_lock_content, pkg_lock_expected", ((NPM_LOCKFILE_JSON, NPM_LOCKFILE_EXPECTED), (None, None)))
@pytest.mark.parametrize("pkg_lock_content, pkg_lock_expected", ((NPM_LOCKFILE_JSON, NPM_LOCKFILE_EXPECTED), (None, None)))
def test_npm_provider(
    config: BaseConfig,
    chdir: Path,
    pkg_lock_content: str,
    pkg_lock_expected: str,
    pkg_shrinkwrap_content: str,
    pkg_shrinkwrap_expected: str,
):
    pkg = chdir / "package.json"
```

**Review by Lee-W — COMMENTED** [review]
Hi @davidlday , thanks for your contribution! The overall implementation is great! Just a few minor thing needs to be fixed

**Review by woile — APPROVED** [review]
LGTM, thank you so much!

File: tests/test_version_providers.py [hunk 6]
```diff
@@ -225,3 +225,98 @@ def test_scm_provider_default_without_commits_and_tags(config: BaseConfig):
...
+        (NPM_LOCKFILE_JSON, NPM_LOCKFILE_EXPECTED, None, None),
+        (None, None, NPM_LOCKFILE_JSON, NPM_LOCKFILE_EXPECTED),
+        (
+            NPM_LOCKFILE_JSON,
+            NPM_LOCKFILE_EXPECTED,
+            NPM_LOCKFILE_JSON,
+            NPM_LOCKFILE_EXPECTED,
+        ),
+    ),
+)
```
Comment (Lee-W) [inline tests/test_version_providers.py:294]:
```suggestion
@pytest.mark.parametrize(
    "pkg_shrinkwrap_content, pkg_shrinkwrap_expected",
    (
        (NPM_LOCKFILE_JSON, NPM_LOCKFILE_EXPECTED), 
        (None, None)
    )
)
@pytest.mark.parametrize(
    "pkg_lock_content, pkg_lock_expected",
    (
        (NPM_LOCKFILE_JSON, NPM_LOCKFILE_EXPECTED), 
        (None, None)
    )
)
```

**Review by Lee-W — APPROVED** [review]
@davidlday Awesome! Just left one minor suggestion. After rebasing, we're good to go

File: tests/test_version_providers.py (hunk 6 above)
Comment (davidlday) [reply tests/test_version_providers.py:294]: @Lee-W  - Applying suggestion now. Sorry, you made this suggestion earlier, and I wasn't getting it. Makes sense to me now.

**Comment by Lee-W** [comment]
Hi @davidlday ,  could you please rebase from the master branch so that we can merge it? You can do it through `git rebase -i master` Thanks!

**Comment by davidlday** [comment]
@Lee-W - apologies, been awhile since I've done a rebase. Let me know if I messed anything up.

**Comment by Lee-W** [comment]
> @Lee-W - apologies, been awhile since I've done a rebase. Let me know if I messed anything up.

No worry  🙂 Just want to fix a CI issue along with this PR.

**Pull request merged**